### DISCUSSION TITLE
Supabase JS v2 Types + React Query

### DISCUSSION THREAD

**Discussion by drewbietron** [supabase/supabase#9530] [Questions]
I'm loving the new type improvements with the v2 JS library ❤️.  I am trying to integrate them with React Query in addition to the React Context examples that have been provided by Supabase.  My trouble has been with getting the types of the client to flow through the context provider.

If I use React Query straight out of the box, without trying to abstract it into a context provider, the types are working since React Query infers the return type of the function that's given.

For example
```typescript
import { useQuery } from "react-query";
import client from './supabase-client';

const results = useQuery("query-key", async () => client.from("users").select("id, name").single());
```
The above example will give me type safety and autocompletion for React Query + Supabase client types.  
```typescript
results.data.data.column -> Property 'column' does not exist on type '{ id: string; name: string; }'
```

What I am trying to do is abstract this function into a context provider so that I can use the same method of fetching data from Supabase via React Query all over the place.  What is becoming tricky is since Context types are not inferred, I have to explicitly declare the return type of the `useSupabaseQuery` function in the provider.  [This is a good example of this being done with the v1 client ](https://github.com/HermanNygaard/supabase-query/blob/main/src/index.tsx#L46)

I am able to get the client typed (since that is just the new SupabaseClient<Database>) but the inference of the Supabase query is what is tricky to type out for the context types.  I have tried as many combinations of the PostgrestJS types I can come up with.  But since the Supabase client infers the methods and types based on what you're typing (.select, .upsert), views vs tables vs storage etc.  I am a bit lost at how to type this out explicitly as the return type of the `useQuery` function within the provider.

Here's what I have so far
```typescript
import { useQuery } from "react-query";

// Provider (abbreviated as this and the context types is where I need help so dont take this too literally 🙂)
interface ContextProps {
  useSupabaseQuery(
    queryCreator: QueryCreator,
    options?: UseQueryOptions & {
      queryKey?: string | unknown[];
    }
  );;
}

const SupabaseContext = createContext<ContextProps>>({});

type QueryCreator = (supabase: SupabaseClient<Database>) => PostgrestFilterBuilder<any, any> // Obv not any, any -- but what 🤯 ??;
function useSupabaseQuery(queryCreator: QueryCreator, options) {
  const query = queryCreator(client);

  function execute(query) {
    new Promise(async (resolve, reject) => {
    const { data, error } = await query;
    if (data) {
      resolve(data);
    } else {
      reject(error);
    }
  });
}

  return useReactQuery(options?.queryKey ?? [query], () => execute(query), options);
}

// Component
const { useSupabaseQuery } = useSupabaseContext();

const results = useSupabaseQuery((client) => client.from("users").select("id, name").single());
// client is typed ^, .select is not //
results.data. // no types returned
```

Been messing with this for a few days now and it would be really awesome to figure out a way to get this within a context so I can have the benefits of React Query's fetching helpers, backed with the type safety of Supabase.  Eventually I'd like to have the query keys read from the Supabase query itself but I gotta crawl first!

Any help is appreciated! 🙏

**Comment by 413n** [comment]
@drewbietron did you got some updates? or did you find a way to do it?

**Reply by drewbietron** [reply]
> @drewbietron did you got some updates? or did you find a way to do it?

Not yet 😕. Still working on it though and have some ideas so will get back to you soon!

**Reply by drewbietron** [reply]
@413n - Here's what I'm current working with.  You still need to pass in a generic to `useQuery`, but it gives you the type safety with your queries, and the type safety from the returned value from React Query.  Would love to take this a little bit further.

```typescript
const query = client
    .from("table")
    .select("id")
    .single();

const res = useQuery<Awaited<typeof query>>(() => query);

res.data.data.id 👍 
res.data.data.foo
// Property 'foo' does not exist on type '{ id: string; } 👍 
```

Caveats are that your types supplied to the useQuery generic could be a lie.

**Reply by 413n** [reply]
Good job there @drewbietron !
Yeah I tried as well to draft a solution, but it's pretty complicated to get the infer type from the param. And I'm not that great at typescript to know how to do it, so I'm counting on you!
Personally I would like to have a solution that include nice handling of errors (according to react-query), the direct return of the supabase `data` in the `useQuery's data return` and the type infer (the most annoying). I will have to take some time to think and retry.
In the meantime keep me updated that I'm really curious about your findings!

**Reply by drewbietron** [reply]
@413n - Some smarter people than me are tackling this exact problem and have done some awesome work.  It already works great with SWR, but for the React Query fans, they're on it too!  

There's a great bit in that repo that [creates a query key for you](https://github.com/psteinroe/supabase-cache-helpers/tree/main/packages/postgrest-filter#postgrestqueryparser-and-postgrestparser), so you can get much more juice out of React Query than what I posted above.  

You can follow along here - https://github.com/psteinroe/supabase-cache-helpers/issues/57

**Reply by 413n** [reply]
That's awesome! Thank you for sharing. Much appreciated. I will follow that issue then.

**Reply by HermanNygaard** [reply]
Hey @drewbietron @413n , I have released a new version that supports Supabase v2 types: https://github.com/HermanNygaard/supabase-query
